File: frontend/src/pages/settings/notificationPrefs.ts
import type { ParseKeys } from 'i18next'
import type { components } from '../../api/generated/schema'

/**
 * The Notifications section's whole logic (#1047), kept out of the component.
 *
 * The repo's harness is `renderToStaticMarkup` with no DOM and no effects, so
 * a rule that lives inside a click handler is a rule nothing can assert. Every
 * question this card answers — which cell is lit, what a click writes, whether
 * the master row's thumb shows — is a pure function of the fetched state and
 * is answered here. Same posture as `updatesFilters.ts` and
 * `requestsQueueCursor.ts`.
 *
 * ## Two switches per row, and they are not the same kind of thing
 *
 * `page` ("show on Updates") is **live behaviour**: the server drops the type
 * out of the feed, the tab counts and the sidebar panel together
 * (`services/notification_prefs.muted_feed_types`).
 *
 * `email` ("email me") is **stored intent only**. Nothing in `backend/` sends
 * email — no provider, no sender, no queue, no template layer. The switch is
 * real, the value is stored, and #2164 honours it when the channel goes live;
 * the card says so in one line of copy, which is the only thing that makes a
 * switch for an absent channel honest rather than the false affordance #1263
 * named. Do not add a sender to make this "work".
 *
 * ## `locked` comes down the wire and is never derived here
 *
 * The rule behind it is requests-section membership, and that set is derived
 * from `FEED_SOURCES` server-side. A second derivation on this side is exactly
 * how a tenth request type gets silently unlocked, so the client is told.
 */

type NotificationPrefOut = components['schemas']['NotificationPrefOut']

export type PrefAxis = 'page' | 'email'

/** The server's answer: every row, resolved, keyed by event. */
export type NotificationPrefs = Readonly<Record<string, NotificationPrefOut>>

export interface NotificationRow {
  /** The event key — matches the server's registry and the stored JSONB key. */
  readonly key: string
  readonly titleKey: ParseKeys<'common'>
  readonly helpKey: ParseKeys<'common'>
}

/**
 * The nine rows, in the order the card draws them, with their copy.
 *
 * Catalog keys are written out rather than built from `key`. A
 * template-literal key typechecks against nothing and is invisible to a locale
 * grep — the same rule `SETTINGS_SECTIONS` and `CookiesSection` already keep.
 *
 * The ORDER is this list's; the VALUES are the server's. A row the server does
 * not know is not drawn (see {@link rowsFor}) rather than drawn with invented
 * state — `__tests__/notificationPrefs.test.tsx` pins the two lists together
 * against the Python registry so that can never happen silently.
 */
export const NOTIFICATION_ROWS: readonly NotificationRow[] = [
  {
    key: 'duel_challenge',
    titleKey: 'settings.notifications.rows.duelChallenge',
    helpKey: 'settings.notifications.rows.duelChallengeHelp',
  },
  {
    key: 'collab_invite',
    titleKey: 'settings.notifications.rows.collabInvite',
    helpKey: 'settings.notifications.rows.collabInviteHelp',
  },
  {
    key: 'invitation_letter',
    titleKey: 'settings.notifications.rows.invitationLetter',
    helpKey: 'settings.notifications.rows.invitationLetterHelp',
  },
  {
    key: 'comment_on_mine',
    titleKey: 'settings.notifications.rows.commentOnMine',
    helpKey: 'settings.notifications.rows.commentOnMineHelp',
  },
  {
    key: 'comment_mention',
    titleKey: 'settings.notifications.rows.commentMention',
    helpKey: 'settings.notifications.rows.commentMentionHelp',
  },
  {
    key: 'vote_on_mine',
    titleKey: 'settings.notifications.rows.voteOnMine',
    helpKey: 'settings.notifications.rows.voteOnMineHelp',
  },
  {
    key: 'level_up',
    titleKey: 'settings.notifications.rows.levelUp',
    helpKey: 'settings.notifications.rows.levelUpHelp',
  },
  {
    key: 'era_announcement',
    titleKey: 'settings.notifications.rows.eraAnnouncement',
    helpKey: 'settings.notifications.rows.eraAnnouncementHelp',
  },
  {
    key: 'global_task',
    titleKey: 'settings.notifications.rows.globalTask',
    helpKey: 'settings.notifications.rows.globalTaskHelp',
  },
]

/** The rows the server sent state for, in this file's order. */
export function rowsFor(prefs: NotificationPrefs): NotificationRow[] {
  return NOTIFICATION_ROWS.filter((row) => prefs[row.key] !== undefined)
}

/**
 * Whether a cell is the reader's to move.
 *
 * Only the PAGE axis can lock. Email is free on every row, including the three
 * requests rows — that is the entire reason the exclusive three-segment control
 * was replaced: *"you should be able to turn off email notifications on
 * collaboration"* while the invite still lands on your Updates page.
 */
export function isLocked(pref: NotificationPrefOut, axis: PrefAxis): boolean {
  return axis === 'page' && pref.locked
}

/**
 * Why a locked page switch cannot move, as a catalog key.
 *
 * Two locks, two reasons, and both fall out of the value the server already
 * sent rather than a third table:
 *
 * - **locked ON** — the row is a request. Something is waiting on an answer
 *   from you and the Requests queue is its only home (ADR-0070), so it may
 *   never be suppressed (owner ruling, 2026-08-19).
 * - **locked OFF** — the row has no Updates row to show. `level_up` is a
 *   self-notification with no feed source: `LevelUpPopup` already tells you at
 *   the moment it happens (owner ruling, 2026-08-18, "no feed row").
 *
 * Locked and *explaining itself* rather than absent: an absent control says
 * nothing, and this page already ships the locked-switch idiom deliberately on
 * the Cookies card for exactly that reason.
 */
export function lockNoteKey(pref: NotificationPrefOut): ParseKeys<'common'> | undefined {
  if (!pref.locked) return undefined
  return pref.page
    ? 'settings.notifications.lockedRequest'
    : 'settings.notifications.lockedNoFeedRow'
}

/**
 * The rows the master row's cell for this axis actually governs.
 *
 * NOT always nine, and it cannot be. The master row is a bulk-set control, and
 * a cell that claimed to set a locked row would either lie or do nothing — so
 * it governs the rows it can set: all nine on `email`, and on `page` the five
 * whose switch is the reader's. This is the one place the 2026-08-19 ruling's
 * "writes it to all nine rows" is read as "all nine it can write", because the
 * same ruling introduced the locks that make the literal reading impossible.
 */
export function governedRows(prefs: NotificationPrefs, axis: PrefAxis): NotificationRow[] {
  return rowsFor(prefs).filter((row) => !isLocked(prefs[row.key], axis))
}

/**
 * The master cell's thumb: **on only when every row it governs is on.**
 *
 * An empty set is not agreement — with nothing to govern the cell has nothing
 * to say, so it reads off rather than reading "all of nothing is true".
 */
export function masterChecked(prefs: NotificationPrefs, axis: PrefAxis): boolean {
  const governed = governedRows(prefs, axis)
  return governed.length > 0 && governed.every((row) => prefs[row.key][axis])
}

/** Flip one cell. Locked cells return the state unchanged — the switch is
 *  already unclickable, and this is the same answer for anything that is not. */
export function toggleCell(
  prefs: NotificationPrefs,
  key: string,
  axis: PrefAxis,
): NotificationPrefs {
  const pref = prefs[key]
  if (!pref || isLocked(pref, axis)) return prefs
  return { ...prefs, [key]: { ...pref, [axis]: !pref[axis] } }
}

/**
 * The master row's click: write one value into every row the cell governs.
 *
 * The value is the OPPOSITE of the thumb, so the control is a switch rather
 * than two buttons — "all on" when any row is off, "all off" when none is.
 * Nothing persists for the master row itself; it has no stored value and never
 * gains one (owner ruling, 2026-08-18).
 */
export function setAll(prefs: NotificationPrefs, axis: PrefAxis): NotificationPrefs {
  const value = !masterChecked(prefs, axis)
  const next = { ...prefs }
  for (const row of governedRows(prefs, axis)) {
    next[row.key] = { ...prefs[row.key], [axis]: value }
  }
  return next
}

/**
 * The PUT body. Every known row, every time.
 *
 * No diffing against what was fetched: the server drops unknown keys and pins
 * a locked row's `page` itself, so a whole-state save is nine small objects
 * and cannot go stale — and a diff would be a second model of "what changed"
 * for a card whose one write is already the whole card.
 */
export function saveBody(prefs: NotificationPrefs): {
  events: Record<string, { page: boolean; email: boolean }>
} {
  const events: Record<string, { page: boolean; email: boolean }> = {}
  for (const row of rowsFor(prefs)) {
    events[row.key] = { page: prefs[row.key].page, email: prefs[row.key].email }
  }
  return { events }
}

File: backend/services/notification_prefs.py
"""Per-ACCOUNT notification preferences (#1047) — the registry and its resolver.

Nine rows, two independent switches each (owner ruling 2026-08-19, which
supersedes #2153's exclusive three-segment control):

``page``
    "Show on Updates" — whether this event may appear in the activity feed.
    **Wired and honoured for real** (owner ruling 2026-08-31):
    :func:`muted_feed_types` feeds ``activity_feed._visible_types``, so the
    list, the tab badges and the sidebar panel all drop the type together.

``email``
    "Email me" — **stored intent only.** Nothing in ``backend/`` sends email:
    no provider, no sender, no queue, no template layer. This value is written
    and read back and does nothing else; #2164 honours it when the channel
    goes live, and the settings card says so in copy. Do NOT add a sender here.

WHY ONE JSONB COLUMN AND NOT A TABLE
------------------------------------
``account.notification_prefs`` is one JSONB, ``server_default '{}'``. Nothing
queries across accounts — a send already knows its recipient — so there is no
join to make cheap and no index to want. **Unset keys resolve to their default
here**, which is what makes adding a tenth event a code change with no
migration and no backfill: an account still holding ``{}`` resolves to every
default, including defaults invented after it was written.

THE LOCKS ARE DERIVED, NEVER LISTED
-----------------------------------
Two rows lock, for two different reasons, and both fall out of the registry:

1. **A requests row may never be suppressed** (owner ruling 2026-08-19). Its
   page switch is locked ON. Membership comes from
   ``activity_feed.REQUEST_ITEM_TYPES``, which is itself derived from
   ``FEED_SOURCES``, so a *tenth* request type cannot be silently unlocked by
   anyone adding it to ``FILTER_REQUESTS``. The rule is requests-section
   membership, not "does it involve me": ``comment_mention`` and
   ``comment_on_mine`` are deliberately not requests and stay disableable.

2. **A row with no feed type has nothing to show.** See ``LEVEL_UP`` below.

``awaiting_submission`` gets no row at all. It is the fourth requests type but
it is *state* rather than an event — the backend refuses to archive it — so
there is nothing to prefer.
"""

from dataclasses import dataclass
from typing import Any, Mapping

from services.activity_feed import (
    FEED_ITEM_TYPE_COLLAB_INVITE,
    FEED_ITEM_TYPE_COMMENT_MENTION,
    FEED_ITEM_TYPE_COMMENT_ON_MINE,
    FEED_ITEM_TYPE_DUEL_CHALLENGE,
    FEED_ITEM_TYPE_ERA_ANNOUNCEMENT,
    FEED_ITEM_TYPE_GLOBAL_TASK,
    FEED_ITEM_TYPE_INVITATION_LETTER,
    FEED_ITEM_TYPE_VOTE_CHANGED_ON_MINE,
    FEED_ITEM_TYPE_VOTE_ON_MINE,
    REQUEST_ITEM_TYPES,
)

#: The one row that is not backed by a feed type. ``level_up`` exists today
#: only as a *taunt* trigger aimed at other people; the owner ruled on
#: 2026-08-18 that it gets **no feed row**, because ``LevelUpPopup`` already
#: tells you at the moment it happens. So its page switch has nothing to
#: filter and renders locked OFF with the reason on the row — the same
#: locked-switch idiom the Privacy card and the three requests rows use,
#: rather than a live switch that would move and change nothing (the
#: false-affordance class #1263 named and the 2026-07-30 ruling forbids).
#: Its **email** switch is the whole point of the row and is fully free.
LEVEL_UP = "level_up"


@dataclass(frozen=True)
class NotificationEvent:
    """One row of the Notifications card.

    ``key`` is the stored JSONB key and the client's copy key — a stable
    identifier, not a feed type, because one row can own more than one type.
    """

    key: str
    #: Feed types the page switch filters. Empty means the row has no feed
    #: presence at all; see :data:`LEVEL_UP`.
    feed_types: tuple[str, ...]
    #: "Email me" starts on for the five things aimed at a person and off for
    #: the four that are ambient or high-frequency (owner ruling 2026-08-18):
    #: a praxis doing well is a dozen mails in an evening, the level-up popup
    #: already says it, and an era announcement goes to everybody at once.
    email_default: bool

    @property
    def page_locked(self) -> bool:
        """Whether "show on Updates" is the reader's to change."""
        if not self.feed_types:
            return True  # nothing to show — locked OFF
        return bool(set(self.feed_types) & REQUEST_ITEM_TYPES)  # locked ON

    @property
    def page_default(self) -> bool:
        """Also the pinned value of a locked row: ON if it has a feed type."""
        return bool(self.feed_types)


#: The nine rows, in the order the card draws them.
NOTIFICATION_EVENTS: tuple[NotificationEvent, ...] = (
    NotificationEvent(
        key=FEED_ITEM_TYPE_DUEL_CHALLENGE,
        feed_types=(FEED_ITEM_TYPE_DUEL_CHALLENGE,),
        email_default=True,
    ),
    NotificationEvent(
        key=FEED_ITEM_TYPE_COLLAB_INVITE,
        feed_types=(FEED_ITEM_TYPE_COLLAB_INVITE,),
        email_default=True,
    ),
    NotificationEvent(
        key=FEED_ITEM_TYPE_INVITATION_LETTER,
        feed_types=(FEED_ITEM_TYPE_INVITATION_LETTER,),
        email_default=True,
    ),
    NotificationEvent(
        key=FEED_ITEM_TYPE_COMMENT_ON_MINE,
        feed_types=(FEED_ITEM_TYPE_COMMENT_ON_MINE,),
        email_default=True,
    ),
    NotificationEvent(
        key=FEED_ITEM_TYPE_COMMENT_MENTION,
        feed_types=(FEED_ITEM_TYPE_COMMENT_MENTION,),
        email_default=True,
    ),
    # One row, two types: a vote landing and a vote changing are both "someone
    # rated my praxis" to a reader, and splitting them would be a switch about
    # the storage rather than about the event.
    NotificationEvent(
        key=FEED_ITEM_TYPE_VOTE_ON_MINE,
        feed_types=(FEED_ITEM_TYPE_VOTE_ON_MINE, FEED_ITEM_TYPE_VOTE_CHANGED_ON_MINE),
        email_default=False,
    ),
    NotificationEvent(key=LEVEL_UP, feed_types=(), email_default=False),
    NotificationEvent(
        key=FEED_ITEM_TYPE_ERA_ANNOUNCEMENT,
        feed_types=(FEED_ITEM_TYPE_ERA_ANNOUNCEMENT,),
        email_default=False,
    ),
    NotificationEvent(
        key=FEED_ITEM_TYPE_GLOBAL_TASK,
        feed_types=(FEED_ITEM_TYPE_GLOBAL_TASK,),
        email_default=False,
    ),
)

EVENTS_BY_KEY: dict[str, NotificationEvent] = {
    event.key: event for event in NOTIFICATION_EVENTS
}


@dataclass(frozen=True)
class ResolvedPref:
    """One row's two switches, after defaults and locks have been applied."""

    page: bool
    email: bool
    locked: bool


def _stored_bool(stored: Any, field: str, fallback: bool) -> bool:
    """Read one switch out of whatever the column happens to hold.

    Defensive rather than paranoid: the column is JSONB written by this app,
    but it is also the one column a hand-edited row or a rolled-back deploy can
    leave a shape in. Anything that is not a real boolean resolves to the
    default, which is exactly what an unset key does.
    """
    if not isinstance(stored, Mapping):
        return fallback
    value = stored.get(field)
    return value if isinstance(value, bool) else fallback


def resolve_prefs(raw: Any) -> dict[str, ResolvedPref]:
    """Every row's live values — the ONE reader of the stored column.

    An account holding ``{}`` (or ``None``, or junk) gets every default. A
    locked row's page value is pinned here rather than trusted from storage,
    so a stale write from before a type joined the requests section cannot
    suppress an obligation.
    """
    stored = raw if isinstance(raw, Mapping) else {}
    resolved: dict[str, ResolvedPref] = {}
    for event in NOTIFICATION_EVENTS:
        row = stored.get(event.key)
        resolved[event.key] = ResolvedPref(
            page=(
                event.page_default
                if event.page_locked
                else _stored_bool(row, "page", event.page_default)
            ),
            email=_stored_bool(row, "email", event.email_default),
            locked=event.page_locked,
        )
    return resolved


def muted_feed_types(raw: Any) -> frozenset[str]:
    """Feed types this account has turned off on the Updates page.

    A request type can never appear here — :attr:`NotificationEvent.page_locked`
    pins those on, so the bell's number and the Requests queue are untouchable
    by this whole file.
    """
    resolved = resolve_prefs(raw)
    return frozenset(
        feed_type
        for event in NOTIFICATION_EVENTS
        if not resolved[event.key].page
        for feed_type in event.feed_types
    )


def apply_update(raw: Any, incoming: Mapping[str, Mapping[str, bool]]) -> dict:
    """The new column value after a client write. Pure — the caller assigns it.

    The trust boundary, and the whole of the validation:

    * **Unknown keys are dropped.** A hostile client cannot grow this column
      without bound, and a stale client naming a retired event degrades instead
      of 400ing on a preferences save.
    * **A locked row's page value is ignored.** It is not the reader's to set,
      so accepting one would store a value that :func:`resolve_prefs` then
      overrides — a lie in the database.

    Every known key is written on every save, not just the ones that differ
    from their default. The column is nine small objects at worst, and a
    "sparse, defaults-only" store would silently re-flip a reader's explicit
    choice the day a default changes.
    """
    stored = dict(raw) if isinstance(raw, Mapping) else {}
    for key, values in incoming.items():
        event = EVENTS_BY_KEY.get(key)
        if event is None:
            continue
        current = stored.get(key)
        stored[key] = {
            "page": (
                event.page_default
                if event.page_locked
                else _stored_bool(values, "page", _stored_bool(current, "page", event.page_default))
            ),
            "email": _stored_bool(
                values, "email", _stored_bool(current, "email", event.email_default)
            ),
        }
    return stored
